module7/Homework7-start/Homework7/Models/JSONDataLoadingStore.swift

```
@@ -0,0 +1,109 @@
+/// released under various Open-Source licenses. Use of those libraries and
+/// frameworks are governed by their own individual licenses.
+///
+/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+/// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+/// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+/// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+/// THE SOFTWARE.
+
+import Foundation
+
+enum JSONDataLoadingStoreDataState {
+  case notLoaded
+  case loading
+  case loaded
+  case writing
+  case errorLoading
+  case errorWriting
+}
+
+enum JSONDataLoadingStoreError: Error, Equatable {
+    case DataFileNotFound(String)
+}
+
+protocol JSONDataLoadingStore: AnyObject {
+  associatedtype DataType: Codable
+  associatedtype DataContainerType: Codable
+  
+  var dataState: JSONDataLoadingStoreDataState { get set }
+  var bundleJSONURL: URL { get set }
+  var documentsJSONURL: URL { get set }
+  var data: DataType? { get set }
+  
+  func extractDataFromContainer(_ container: DataContainerType) -> DataType?
+  func createContainerFromData(_ data: DataType?) -> DataContainerType
+  
+}
+
+extension JSONDataLoadingStore {
+  
+  func readJSON() {
+    do {
+      dataState = .loading
+      data = try readJSON(with: bundleJSONURL, fallingBackTo: documentsJSONURL)
+      dataState = .loaded
+    } catch {
+      dataState = .errorLoading
+    }
+  }
+
+  func readJSONFromUrl(url: URL) -> DataType? {
+    let decoder = JSONDecoder()
+    do {
+      let unstructuredUserData = try Data(contentsOf: url)
+      let dataJSONContainer = try decoder.decode(DataContainerType.self, from: unstructuredUserData)
+      return extractDataFromContainer(dataJSONContainer)
+    } catch let error {
+      print(error)
```

You'd probably want to mark this with a `throw` and then throw your error instead of printing it, since `readJSON(with:, fallingBackTo:)` throws when it cant find the file, and this one will have an error if the file is corrupted.

---

Good catch!